File: 2023/etc/bs/12015C.py
# O(NlgN) (naive: O(N^2))

import sys
input = lambda: sys.stdin.readline().rstrip()

N = int(input())
A = list(map(int, input().split()))

# bisect.bisect_left()가 더 빠름
LIS = []
lower_bound = None
def bound(l, r, key):
	global lower_bound
	while True:
		if l == r:
			lower_bound = l
			return -1
		
		m = (l + r) // 2
		if key == LIS[m]:
			lower_bound = m
			return m
		else:
			if key < LIS[m]:
				r = m
			else:
				l = m + 1

for i in range(N):
	bound(0, len(LIS), A[i])
	if lower_bound == len(LIS):
		LIS.append(0)
	LIS[lower_bound] = A[i]

print(len(LIS))

File: 2023/etc/bs/12015C.cpp
#include <cstdio>
#include <algorithm>
using namespace std;

int N;
int A[1'000'000];
int LIS[1'000'000], len_LIS;

int main()
{
	scanf("%d", &N);
	for (int i = 0; i < N; i++) {
		int x;
		scanf("%d", &x);

		auto it = lower_bound(LIS, LIS + len_LIS, x);
		if (it == LIS + len_LIS)
			len_LIS += 1;
		*it = x;
	}
	printf("%d", len_LIS);
}